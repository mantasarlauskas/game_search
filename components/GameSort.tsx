import React, { useState } from 'react';
import { Popover } from 'react-tiny-popover';
import styles from 'components/GameSort.module.scss';
import { useRouter } from 'next/router';
import { getOrdering } from 'utils/ordering';

export enum SortMode {
    RELEVANCE = '',
    NAME = 'name',
    DATE_ADDED = 'added',
    RELEASE_DATE = '-released',
    RATING = '-rating'
}

const titles = {
    [SortMode.RELEVANCE]: 'Relevance',
    [SortMode.DATE_ADDED]: 'Date added',
    [SortMode.NAME]: 'Name',
    [SortMode.RELEASE_DATE]: 'Release date',
    [SortMode.RATING]: 'Average rating',
};

const sortOptions = [
    SortMode.RELEVANCE,
    SortMode.DATE_ADDED,
    SortMode.NAME,
    SortMode.RELEASE_DATE,
    SortMode.RATING,
];

function GameSort() {
    const router = useRouter();
    const ordering = getOrdering(router.query);
    const [isOpen, setIsOpen] = useState(false);

    function onOptionClick(value: SortMode) {
        setIsOpen(false);
        if (value !== ordering) {
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            const { ordering: prevOrdering, ...queryParams } = router.query;
            router.push({
                pathname: router.pathname,
                query: value ? {
                    ...queryParams,
                    ordering: value,
                } : queryParams,
            }, undefined, { shallow: true });
        }
    }

    const button = (
        <button
            type="button"
            className={styles.root}
            onClick={() => setIsOpen(!isOpen)}
        >
            Order by:
            <div className={styles.title}>
                {titles[ordering]}
            </div>
        </button>
    );

    if (typeof window === 'undefined') {
        return button;
    }

    return (
        <Popover
            isOpen={isOpen}
            padding={4}
            positions={['bottom']}
            align="start"
            onClickOutside={() => setIsOpen(false)}
            content={(
                <div className={styles.options}>
                    {sortOptions.map((value) => (
                        <button
                            type="button"
                            className={styles.option}
                            key={value}
                            onClick={() => onOptionClick(value)}
                        >
                            {titles[value]}
                        </button>
                    ))}
                </div>
            )}
        >
            {button}
        </Popover>
    );
}

export default GameSort;
