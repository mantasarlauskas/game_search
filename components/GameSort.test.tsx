import { render, screen } from '@testing-library/react';
import GameSort from 'components/GameSort';
import userEvent from '@testing-library/user-event';

const mockPush = jest.fn();
jest.mock('next/router', () => ({
    useRouter() {
        return {
            query: { ordering: 'added' },
            push: mockPush,
        };
    },
}));

describe('<GameSort />', () => {
    it('matches snapshot', () => {
        const { container } = render(<GameSort />);
        expect(container).toMatchSnapshot();
    });

    it('shows correct sort value', () => {
        render(<GameSort />);
        expect(screen.getByText('Date added')).toBeInTheDocument();
    });

    it('changes sort value', () => {
        render(<GameSort />);
        userEvent.click(screen.getByText('Date added'));
        userEvent.click(screen.getByText('Average rating'));
        expect(mockPush).toBeCalledTimes(1);
        expect(mockPush).toBeCalledWith(expect.objectContaining({
            query: { ordering: '-rating' },
        }), undefined, { shallow: true });
    });
});
