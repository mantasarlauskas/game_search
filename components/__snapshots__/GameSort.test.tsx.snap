// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameSort /> matches snapshot 1`] = `
<div>
  <button
    class="root"
    type="button"
  >
    Order by:
    <div
      class="title"
    >
      Date added
    </div>
  </button>
</div>
`;
